content/algorithms/sub_isomorphism/subgraph_iso_img.md

```diff
@@ -0,0 +1,329 @@
+
++++
+
+We will work on $C_\alpha$ networks. 
+
+Let's see graphs of proteins with PDB codes 1CRN(Plant protein), 1FN3(oxygen storage/transport in the human body) and 1EGJ(Plant Protein).
+
+```{code-cell} ipython3
+p_1CRN = nx.read_edgelist("data/1CRN_edgelist.txt", nodetype=str)
+nx.draw_spring(p_1CRN, node_color ="g")
+plt.title("1CRN Plant Protein", fontweight = "bold")
+plt.show()
+```
+
+```{code-cell} ipython3
+p_1FN3 = nx.read_edgelist("data/1FN3_edgelist.txt", nodetype=str)
+nx.draw_spring(p_1FN3, node_color = "r")
+plt.title("1FN3 oxygen storage/transport Protein", fontweight = "bold")
+plt.show()
+```
+
+```{code-cell} ipython3
+p_1EJG = nx.read_edgelist("data/1EJG_edgelist.txt", nodetype=str)
+nx.draw_spring(p_1EJG, node_color = "g")
+plt.title("1EJG Plant Protein", fontweight = "bold")
+plt.show()
+```
+
+We can identify graphlets (induced subgraphs) that are present in these graphs and use that to classify proteins. We can extract some subgraphs from the proteins and test if they are present in other proteins. Let's find a subgraph of 1CRN that is also a subgraph of 1EJG but not of 1FN3. This is interenting because 1CRN and 1EJG are both plant proteins but 1FN3 is not. But clearly to decide if this graphlet is particular of plant proteins we should test it in more proteins.
+
+```{code-cell} ipython3
+#Get a induced subgraph from 1CRN
+graphlet = p_1CRN.subgraph(["A"+str(i) for i in range(0, 30)])
+nx.draw(graphlet)
+
+#Test if the graphlet is a subgraph of each protein 
+g1 = iso.GraphMatcher(p_1CRN, graphlet)
+print("Graphlet present in 1CRN ", g1.subgraph_is_isomorphic())
+
+g2 = iso.GraphMatcher(p_1FN3, graphlet)
+print("Graphlet present in 1FN3" ,g2.subgraph_is_isomorphic())
+
+g3 = iso.GraphMatcher(p_1EJG, graphlet)
+print("Graphlet present in 1EJG", g3.subgraph_is_isomorphic())
+```
+
+On a similar way we can find a graphlet in 1FN3 that is not present in 1CRN and 1EJG.
+
+```{code-cell} ipython3
+#Get a induced subgraph from 1FN3
+graphlet = p_1FN3.subgraph(["A"+str(i) for i in range(100, 110, 1)])
+plt.rcParams["figure.figsize"] = (5,5)
+nx.draw(graphlet)
+
+#Test if the graphlet is a subgraph of each protein 
+g1 = iso.GraphMatcher(p_1CRN, graphlet)
+print("Graphlet present in 1CRN ", g1.subgraph_is_isomorphic())
+
+g2 = iso.GraphMatcher(p_1FN3, graphlet)
+print("Graphlet present in 1FN3" ,g2.subgraph_is_isomorphic())
+
+g3 = iso.GraphMatcher(p_1EJG, graphlet)
+print("Graphlet present in 1EJG", g3.subgraph_is_isomorphic())
+```
+
+Another option is to use randomly generated graphs given a certain number of nodes, for example, using Erdos-graphs generators. *erdos_renyi_graph(n, p, seed)* generates a graph of n nodes in which all possible edges are added with probability p. We can find a random graphs that is present in 1FN3 but not in 1CRN and 1EJG.
```

Comment:
Yes, Mostly useful subgraphs come from the proteins but also random graphs are used to experiment.  I used erdos-reyni because they are easy to understand but they also build random graphs from subgraphs of the proteins, for example, adding new edges randomly.